src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java

```diff
@@ -151,76 +151,76 @@ private static void writeReadsADAM(
         // we are writing the Avro schema to the Configuration as a JSON string. The AvroParquetOutputFormat class knows
         // how to translate objects in the Avro data model to the Parquet primitives that get written.
         AvroParquetOutputFormat.setSchema(job, AlignmentRecord.getClassSchema());
-        deleteHadoopFile(outputFile);
+        deleteHadoopFile(outputFile, ctx.hadoopConfiguration());
         rddAlignmentRecords.saveAsNewAPIHadoopFile(
                 outputFile, Void.class, AlignmentRecord.class, AvroParquetOutputFormat.class, job.getConfiguration());
     }
 
-    private static void writeReadsSharded(
-            final JavaSparkContext ctx, final String outputFile, final JavaRDD<SAMRecord> reads,
-            final SAMFileHeader header) throws IOException {
-        // Set the header on the main thread.
+
+    private static void saveAsShardedHadoopFiles(JavaSparkContext ctx, String outputFile, JavaRDD<SAMRecord> reads, SAMFileHeader header, boolean writeHeader) throws IOException {
+        // Set the static header on the driver thread.
         SparkBAMOutputFormat.setHeader(header);
-        // MyOutputFormat is a static class, so we need to copy the header to each worker then call
-        final Broadcast<SAMFileHeader> broadcast = ctx.broadcast(header);
-        final JavaRDD<SAMRecord> readsRDD = reads.mapPartitions(readIterator -> {
-            SparkBAMOutputFormat.setHeader(broadcast.getValue());
-            return new IteratorIterable<>(readIterator);
-        });
 
-        // The expected format for writing is JavaPairRDD where the key is ignored and the value is a
-        // SAMRecordWritable. We use SAMRecord as the key, so we can sort the reads before writing them to a file.
-        final JavaPairRDD<SAMRecord, SAMRecordWritable> rddSamRecordWriteable = readsRDD.mapToPair(read -> {
-            read.setHeader(broadcast.getValue());
-            final SAMRecordWritable samRecordWritable = new SAMRecordWritable();
-            samRecordWritable.set(read);
-            return new Tuple2<>(read, samRecordWritable);
-        });
+        final Broadcast<SAMFileHeader> headerBroadcast = ctx.broadcast(header);
 
-        deleteHadoopFile(outputFile);
-        rddSamRecordWriteable.saveAsNewAPIHadoopFile(outputFile, SAMRecord.class, SAMRecordWritable.class, SparkBAMOutputFormat.class);
+        // SparkBAMOutputFormat is a static class, so we need to copy the header to each worker then call
+        final JavaRDD<SAMRecord> readsRDD = setHeaderForEachPartition(reads, headerBroadcast);
+
+        // The expected format for writing is JavaPairRDD where the key is ignored and the value is SAMRecordWritable.
+        final JavaPairRDD<Void, SAMRecordWritable> rddSamRecordWriteable = pairReadsWithSAMRecordWritables(headerBroadcast, readsRDD);
+
+        rddSamRecordWriteable.saveAsNewAPIHadoopFile(outputFile, SAMRecord.class, SAMRecordWritable.class, getOutputFormat(writeHeader));
```

Comment:
It was removed because i believe it's redundant with the deletion that happens as part of `SparkBAMOutputFormat.checkOutputSpecs`  

I added 'testWritingToAnExistingFileHDFS` to verify that it still works when an hdfs file already existed.